### ISSUE TITLE
WSLg overwrites DISPLAY variable when it is set

### ISSUE THREAD

**Issue by elibarzilay** [microsoft/wslg#390]
The normal way of setting environment variables on linux should be to respect user settings.  However, WSLg will just overwrite your `DISPLAY` if you've set its value somehow.  In my case, I have some extensive scripts that find the correct IP on startup and registers that in a `DISPLAY` environment variable on the windows side with `WSLENV` making it propagate to the WSL side.  This failed miserably with WSLg overwriting it with `:0`.  (I'd obviously prefer using the WSLg over a third party X server, but just a few quick experiments made it clear to me that it is not nearly robust enough for actual use.)

After a whole bunch of fumbling around, I resorted to `guiApplications=false` in `.wslconfig`, but this is hard to find, it disables wslg for all distros, and it also disables the pulseaudio server.

# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.22000.100]
Your Distribution version: 20.04
Your WSLg version: 1.0.24
```

# Steps to reproduce

In your Windows environment variables:
* Set a `DISPLAY` variable with whatever
* Set `WSLENV` to include it, eg, `Path/ul:DISPLAY`

#  Expected behavior

`DISPLAY` should have the value set on the windows side.

# Actual behavior

`DISPLAY` gets overwritten by `:0`.

**Comment by hideyukn88** [comment]
@elibarzilay, thanks for reporting the issue. If WSLg is enabled, it assumes user is going to use WSLg as it's X server (at least currently, since WSLg doesn't support to co-exist with other X server solution), thus it automatically sets DISPLAY to :0 where WSLg's X server is listening. So would you please clarify your usage why you wants have custom DISPLAY path while WSLg is enabled ? thanks for trying out WSLg!

**Comment by elibarzilay** [comment]
@hideyukn88, such arguments of "you installed X and therefore we're going to change your environment to use it" are, in my experience, *never* a good idea.  To be more specific:

* My problem started from just update my WSL installation.  The process went as usual, no warnings about changing anything for me.  The thing is that WSLg is included in the update and therefore it got turned on and broke my working environment, which is *not* something that I wanted.  And my environment is "showing" what it wants by the fact that I do have `DISPLAY` set, so if only  WSLg would set it *only if it's not already set* then I could save a day of figuring out how to avoid it.

* I have it disabled now so I can work, but I would like it if I could still use the pulseaudio server.  The two are tied together, and since I must disable all of WSLg globally, I cannot use the audio.

* I would have probably set a second linux distribution just to try out the GUI functionality from time to time, but again, this is impossible since I only get a global on-off switch.

* Even more likely, I'd just run WSLg with my usual third-party X server, and try random things with the WSLg server (ie, I'd run an occasional `DISPLAY=:0 xterm`).  Again, I can't do that since it can only be disabled globally.

* There are other legitimate uses of multiple `$DISPLAY` settings, for example redirecting to a different (possibly remote) display, working with a frame-buffer-based server (like `Xvnc`) and more.  Ignoring what users set in their environment is extremely bad for all of this.

**Comment by crramirez** [comment]
I agree with @elibarzilay, independently of the use case if WSLg sees the DISPLAY already set it should not override it. 

The same rule applies in the opposite, if inside the distro there is a script that sets DISPLAY for xserver it should always first check if it is already set (in this case by WSLg)

**Comment by hideyukn88** [comment]
@elibarzilay, @crramirez , thank you very much for detail and valuable feedback. Current WSLg behavior (override DISPLAY) is indented. thus I removed "bug" tag and replace it with "enhancement". We will have internal discussion with team (and of course monitor "upvote" of this issue), thank you very much for trying out WSLg!

**Comment by elibarzilay** [comment]
@hideyukn88, I hope that the title change would not derail the discussion -- since the *main* point is not just to coexist with another X server or partial features, but rather respect for (intentional) user settings, whatever they may be.

**Comment by hideyukn88** [comment]
@elibarzilay, I understand that and, while I personally agree with you, but as a product we will need to maintain certain integrity as a whole solution. 

As you might see, several X11 socket issue has been posted there and we have FAQ for that (https://github.com/microsoft/wslg/wiki/Diagnosing-%22cannot-open-display%22-type-issues-with-WSLg). Mainly this is because user (intentionally or unintentionally) customize it, or left over from other solutions cause conflicts, thus honor the user settings might not always give "just work" experience.

And current design of WSLg is to provide whole packaged solution (Wayland, X11, audio input/output and etc) or otherwise disable entirely. Thus I changed the title to something more reasonable to express the usage scenario for "why we might want to do this (= respects user settings)".

Thank you very much, we always welcome feedback!

**Comment by Masamune3210** [comment]
IMHO considering we are talking about something already as techy as a Linux distro and issue, worrying about adhering to "it just works" might not be the right idea. Chances are people are at least savvy enough to know Google-fu or to know if they have mucked with envvars

**Comment by hideyukn88** [comment]
@Masamune3210, Just to be sure, this only apply to WSLENV. Any env. can be still override by bash or .profile. WSLg don't override what configured in user-distro side.

**Comment by hideyukn88** [comment]
And update the issue title to reflect the precise feedback from issue opener.

**Comment by elibarzilay** [comment]
+1 to @Masamune3210's point -- WSL should generally be a linux experience, so following linux conventions is IMO more important than any notion of "just works".

Re `WSLENV` -- I think that the assumption that this is the only issue here is wrong, even if in practice that's the only way to do it currently.  I was actually surprised that `wsl env` goes first through a shell rather than running `env` directly, and therefore I won't be surprised if this is changed in the future to not go through your shell.

And as a sidenote, another reason to avoid relying on `.profile` is that different shells behave differently.  And in my case of zsh setup it uses different initialization files.

**Comment by elibarzilay** [comment]
(I forgot to include earlier the existence of `wsl --exec` which *won't* use the shell for initialization.)

**Comment by hideyukn88** [comment]
Hi @elibarzilay, thank you very much for your feedback. We've made a change to honor environment variables specified by WSLENV. The change needs to be made outside of WSLg project, thus I can't say for sure when it will be available, but it is coming. Thanks again for trying out WSLg!

**Comment by elibarzilay** [comment]
@hideyukn88, thanks!

**Comment by mangkoran** [comment]
Since this issue has overwritten `XDG_RUNTIME_DIR` discussion in it, I have a question: Is it intended that `XDG_RUNTIME_DIR` is overwritten by WSLg? I ran this command
```
> echo $XDG_RUNTIME_DIR
/mnt/wslg/runtime-dir
```
and it seems the variable is set by WSLg. Is this needed for WSLg to work?

**Comment by hideyukn88** [comment]
This is fixed in WSL from aka.ms/wslstorepage, thanks!

**Comment by hideyukn88** [comment]
@mangkoran, yes, that has to be specified for Wayland compositor to work, thanks!
